command/pipe.go

```diff
@@ -0,0 +1,355 @@
+			Name:        "no-such-upload-retry-count",
+			Usage:       "number of times that a request will be retried on NoSuchUpload error; you should not use this unless you really know what you're doing",
+			DefaultText: "0",
+			Hidden:      true,
+		},
+		&cli.BoolFlag{
+			Name:    "no-clobber",
+			Aliases: []string{"n"},
+			Usage:   "do not overwrite destination if already exists",
+		},
+	}
+	return pipeFlags
+}
+
+func NewPipeCommand() *cli.Command {
+	cmd := &cli.Command{
+		Name:               "pipe",
+		HelpName:           "pipe",
+		Usage:              "upload to remote from stdin",
+		Flags:              NewPipeCommandFlags(),
+		CustomHelpTemplate: pipeHelpTemplate,
+		Before: func(c *cli.Context) error {
+			err := validatePipeCommand(c)
+			if err != nil {
+				printError(commandFromContext(c), c.Command.Name, err)
+			}
+			return err
+		},
+		Action: func(c *cli.Context) (err error) {
+			defer stat.Collect(c.Command.FullName(), &err)()
+
+			pipe, err := NewPipe(c, false)
+			if err != nil {
+				return err
+			}
+			return pipe.Run(c.Context)
+		},
+	}
+
+	cmd.BashComplete = getBashCompleteFn(cmd, false, false)
+	return cmd
+}
+
+// Pipe holds pipe operation flags and states.
+type Pipe struct {
+	dst         *url.URL
+	op          string
+	fullCommand string
+
+	deleteSource bool
+
+	// flags
+	noClobber        bool
+	storageClass     storage.StorageClass
+	encryptionMethod string
+	encryptionKeyID  string
+	acl              string
+	cacheControl     string
+	expires          string
+	contentType      string
+	contentEncoding  string
+
+	// region settings
+	dstRegion string
+
+	// s3 options
+	concurrency int
+	partSize    int64
+	storageOpts storage.Options
+}
+
+// NewPipe creates Pipe from cli.Context.
+func NewPipe(c *cli.Context, deleteSource bool) (*Pipe, error) {
+	fullCommand := commandFromContext(c)
+
+	dst, err := url.New(c.Args().Get(0),
+		url.WithRaw(c.Bool("raw")))
+	if err != nil {
+		printError(fullCommand, c.Command.Name, err)
+		return nil, err
+	}
+
+	return &Pipe{
+		dst:          dst,
+		op:           c.Command.Name,
+		fullCommand:  fullCommand,
+		deleteSource: deleteSource,
+		// flags
+		noClobber:        c.Bool("no-clobber"),
+		storageClass:     storage.StorageClass(c.String("storage-class")),
+		concurrency:      c.Int("concurrency"),
+		partSize:         c.Int64("part-size") * megabytes,
+		encryptionMethod: c.String("sse"),
+		encryptionKeyID:  c.String("sse-kms-key-id"),
+		acl:              c.String("acl"),
+		cacheControl:     c.String("cache-control"),
+		expires:          c.String("expires"),
+		contentType:      c.String("content-type"),
+		contentEncoding:  c.String("content-encoding"),
+		dstRegion:        c.String("destination-region"),
+
+		// s3 options
+		storageOpts: NewStorageOpts(c),
+	}, nil
+}
+
+// Run starts copying stdin output to destination.
+func (c Pipe) Run(ctx context.Context) error {
+	var task = c.prepareUploadTask(ctx, c.dst)
+	var err = task()
+
+	return err
+}
```

Comment:
This function is essentially 1 line if you take a close look:

`return c.prepareUploadTask(ctx, c.dst)()`

`prepareUploadTask` and `doUpload` helps `cp` command readable but in here, they don't add much. Let's `prepareUploadTask` and `doUpload` code into `Run` method.

Reply:
Resolved. https://github.com/peak/s5cmd/pull/587/commits/b75cde65e62a8a549f7b618cc6b99f99d7f4d345